src/Transport/Sending/DefaultOutgoingBatchRouter.cs

```diff
@@ -33,21 +33,24 @@ public DefaultOutgoingBatchRouter(IConvertOutgoingMessagesToBrokeredMessages out
             maximuMessageSizeInKilobytes = settings.Get<int>(WellKnownConfigurationKeys.Connectivity.MessageSenders.MaximumMessageSizeInKilobytes);
         }
 
-        public async Task RouteBatches(IEnumerable<Batch> outgoingBatches, ReceiveContext context)
+        public Task RouteBatches(IEnumerable<Batch> outgoingBatches, ReceiveContext context)
         {
+            var pendingBatches = new List<Task>();
             foreach (var batch in outgoingBatches)
             {
-                await RouteBatch(batch, context).ConfigureAwait(false);
+                pendingBatches.Add(RouteBatch(batch, context));
             }
+            return Task.WhenAll(pendingBatches);
         }
 
-        public async Task RouteBatch(Batch batch, ReceiveContext context)
+        public Task RouteBatch(Batch batch, ReceiveContext context)
         {
             var outgoingBatches = batch.Operations;
-            if (!outgoingBatches.Any()) return;
```

Comment:
I'm not entirely on board with this one. We remove the check for optimization. But we continue to execute the rest of the code. How much of an optimization is that anyway? Wouldn't exiting early and _not_ performing any other operations be better?